pkg/sql/sem/tree/constant.go

```diff
@@ -369,12 +369,14 @@ func (expr *NumVal) ResolveAsType(
 }
 
 func intersectTypeSlices(xs, ys []*types.T) (out []*types.T) {
+	seen := make(map[*types.T]bool)
```

Comment:
nit: use `map[*types.T]struct{}` instead, with `_, ok := seen[x]` / `seen[x] = struct{}{}`

also this works on a map of pointers -- can we have the same type represented as two different pointers?

would it better to just have a test that `StrValAvailAllParsable` et al have no duplicates?

Reply:
> nit: use `map[*types.T]struct{}` instead, with `_, ok := seen[x]` / `seen[x] = struct{}{}`

done

> can we have the same type represented as two different pointers?
i don't think the same type can be represented as two different pointers. or rather, they definitely could, but the assumption everywhere in the codebase (i believe) is that we only operate on the pointers. i'm basing this on #48279. cc @yuzefovich: curious if you have a more well-thought out answer to oliver's question than i do.

> would it better to just have a test that StrValAvailAllParsable et al have no duplicates?

i thought about that.. but didn't do it since i felt like there probably are many slices in the code that need to be sets. i wanted to have a discussion about whether we should be testing for set-hood more generally. or maybe linting? i'm not sure where to have that discussion, but in the mean time i def can add the test you asked for